gnpy/core/convert.py

```diff
@@ -511,7 +512,11 @@ def parse_excel(input_filename):
         raise ValueError(f'Duplicate city: {all_cities}')
     if any(ln.from_city not in all_cities or
            ln.to_city   not in all_cities for ln in links):
-        raise ValueError(f'Bad link.')
+        bad_links = []
+        for lnk in links:
+            if lnk.from_city not in all_cities or lnk.to_city not in all_cities:
+                bad_links.append([lnk.from_city, lnk.to_city])
```

Comment:
Is this better like this ?

Reply:
The code checks all the links twice (once in the `any(...)`, then once more in the explicit loop). The benefit of this is that `bad_links` can be scoped. I don't really care that much, but if I was writing this, I would probably do it like this:

```py
bad_links = []
for link in links:
  if ...:
    bad_links.append((link.from_city, link.to_city))
if not bad_links.empty():
  raise NetworkTopologyError(...)
```

Feel free to approve and merge as-is if you prefer the existing patch.

Reply:
understood , thanks